### ISSUE TITLE
[webview_flutter] Add example file chooser callback to README

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#118836]
https://github.com/flutter/plugins/pull/6881 added the ability to get a callback for file `<input>`s. This issue tracks adding sample code to the `webview_flutter_android` readme for implementing the callback using `file_selector_android`, once that package lands.

**Comment by AKortak** [comment]
Does this also support uploading other files?

**Comment by IPRamaAnadya** [comment]
@tp im sorry sir, i already tried but still doesnt work.
can u help me?

```
  //home.controller.dart

  late WebViewController controller;
  late PlatformWebViewControllerCreationParams params;
  final Rx<AppState> state = AppState.none.obs;

  @override
  void onInit() async {

    state.value = AppState.loading;
    update();

    if (WebViewPlatform.instance is WebKitWebViewPlatform) {
      params = WebKitWebViewControllerCreationParams(
        allowsInlineMediaPlayback: true,
        mediaTypesRequiringUserAction: const <PlaybackMediaTypes>{},
      );

    } else {
      params = AndroidWebViewControllerCreationParams();
    }

    controller = WebViewController.fromPlatformCreationParams(params)
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setBackgroundColor(const Color(0x00000000))
      ..setNavigationDelegate(
        NavigationDelegate(
          onProgress: (int progress) async {},
          onPageStarted: (String url) {
            state.value = AppState.loading;
            update();
          },
          onPageFinished: (String url) async {
          },
          onWebResourceError: (WebResourceError error) {},
          onNavigationRequest: (NavigationRequest request) {},
        ),
      )
      ..loadRequest(Uri.parse('https://www.example.com'))
      ..addJavaScriptChannel("name", onMessageReceived: (message) {});

    if (controller.platform is WebKitWebViewController) {
      (controller.platform as WebKitWebViewController)
          .setAllowsBackForwardNavigationGestures(true);
    }

    if(controller.platform is AndroidWebViewController) {
      final androidController = controller.platform as AndroidWebViewController;
      await androidController.setOnShowFileSelector(_androidFilePicker);
    }

    super.onInit();
  }
```

**Comment by IPRamaAnadya** [comment]
> Does this also support uploading other files?
> 
> I tried it and it does. Thanks alot for adding the setOnShowFileSelector() method.

Sure, 
to uploading other files, you can use ```file_picker: any``` and add new desicion for filetype in ```_androidFilePicker``` method.

this is my code for it:
```
Future<List<String>> _androidFilePicker(FileSelectorParams params) async {

    if (params.acceptTypes.any((type) => type == 'image/*')) {
      final picker = image_picker.ImagePicker();
      final photo = await picker.pickImage(source: image_picker.ImageSource.camera);

      if (photo == null) {
        return [];
      }

      final imageData = await photo.readAsBytes();
      final decodedImage = image.decodeImage(imageData)!;
      final scaledImage = image.copyResize(decodedImage, width: 500);
      final jpg = image.encodeJpg(scaledImage, quality: 90);

      final filePath = (await getTemporaryDirectory()).uri.resolve(
        './image_${DateTime.now().microsecondsSinceEpoch}.jpg',
      );
      final file = await File.fromUri(filePath).create(recursive: true);
      await file.writeAsBytes(jpg, flush: true);

      return [file.uri.toString()];
    }
    else if (params.acceptTypes.any((type) => type == 'application/*')) {
      FilePickerResult? result = await FilePicker.platform.pickFiles(allowMultiple: true);

      if (result != null) {
        List<File> files = result.paths.map((path) => File(path!)).toList();
        return files.map((file) => file.uri.toString()).toList(growable: false);
      }
    }
    return [];
  }
```

**Comment by Gaurav-CareMonitor** [comment]
Used this for picking single and multiple files

 initFilePicker() async {
    if (Platform.isAndroid) {
      final androidController = (controller.platform
          as webview_flutter_android.AndroidWebViewController);
      await androidController.setOnShowFileSelector(_androidFilePicker);
    }
  }

  /// This method is called when the user tries to upload a file from the webview.
  /// It will open the file picker and return the selected files.
  /// If the user cancels the file picker, it will return an empty list.
  ///
  /// Returns uri's of the selected files.
  Future<List<String>> _androidFilePicker(
      webview_flutter_android.FileSelectorParams params) async {
    try {
      if (params.mode ==
          webview_flutter_android.FileSelectorMode.openMultiple) {
        final attachments =
            await FilePicker.platform.pickFiles(allowMultiple: true);
        if (attachments == null) return [];

        return attachments.files
            .where((element) => element.path != null)
            .map((e) => File(e.path!).uri.toString())
            .toList();
      } else {
        final attachment = await FilePicker.platform.pickFiles();
        if (attachment == null) return [];
        File file = File(attachment.files.single.path!);
        return [file.uri.toString()];
      }
    } catch (e, s) {
    
      return [];
    }
  }

**Comment by stuartmorgan** [comment]
For any further discussion of alternate solutions, please link to another forum, such as a Stack Overflow question, and post suggestions or questions there. This issue is just tracking what is described in the original issue description.

See https://github.com/flutter/flutter/wiki/Issue-hygiene#comments-providing-workarounds

**Comment by kohj1018** [comment]
@tp
How did you pass the controller created in the code below to the webview?
```dart
if (Platform.isAndroid) {
  final controller = (_webViewController.platform as webview_flutter_android.AndroidWebViewController);
  await controller.setOnShowFileSelector(_androidFilePicker);
}
```

The webview of webview_flutter only receives the controller as the WebViewController type, causing problems like the picture below.
![image](https://user-images.githubusercontent.com/68740465/222493294-323a5fd9-b689-45a8-a123-457fa76c5475.png)

Can you give me additional information?
Thank you very much for the good example code.

**Comment by followthemoney1** [comment]
@Gaurav-CareMonitor you saved my life

**Comment by 0x1af2aec8f957** [comment]
> For anyone else who might stumble across this ticket while wanting to support file uploads in web views on Android, here's our code we used to add capturing camera pictures to upload via the embedded web page:
> 
> Packages used:
> 
> ```dart
> import 'package:image/image.dart' as image;
> import 'package:image_picker/image_picker.dart' as image_picker;
> import 'package:webview_flutter_android/webview_flutter_android.dart' as webview_flutter_android;
> ```
> 
> Then after creating the web view controller, attach a custom file upload handler for Android:
> 
> ```dart
> _webViewController = WebViewController();
> …
> 
> if (Platform.isAndroid) {
>   final controller = (_webViewController.platform as webview_flutter_android.AndroidWebViewController);
>   await controller.setOnShowFileSelector(_androidFilePicker);
> }
> ```
> 
> For our case we only need to support camera uploads, so the following suffices:
> 
> ```dart
> Future<List<String>> _androidFilePicker(webview_flutter_android.FileSelectorParams params) async {
>   if (params.acceptTypes.any((type) => type == 'image/*')) {
>     final picker = image_picker.ImagePicker();
>     final photo = await picker.pickImage(source: image_picker.ImageSource.camera);
> 
>     if (photo == null) {
>       return [];
>     }
> 
>     final imageData = await photo.readAsBytes();
>     final decodedImage = image.decodeImage(imageData)!;
>     final scaledImage = image.copyResize(decodedImage, width: 500);
>     final jpg = image.encodeJpg(scaledImage, quality: 90);
> 
>     final filePath = (await getTemporaryDirectory()).uri.resolve(
>           './image_${DateTime.now().microsecondsSinceEpoch}.jpg',
>         );
>     final file = await File.fromUri(filePath).create(recursive: true);
>     await file.writeAsBytes(jpg, flush: true);
> 
>     return [file.uri.toString()];
>   }
> 
>   return [];
> }
> ```

`image_picker` now supports directly obtaining the `file path`, and the path should be returned directly to the `webview` instead of continuing to create the file, which can avoid the screen being in a black screen for a long time when creating a large file :

```dart
Future<List<String>> _androidFilePicker(webview_flutter_android.FileSelectorParams params) async {
  if (params.acceptTypes.any((type) => type == 'image/*')) {
    final picker = image_picker.ImagePicker();
    final photo = await picker.pickImage(source: image_picker.ImageSource.camera);

    if (photo == null) {
      return [];
    }

    return [Uri.file(photo.path).toString()];
  }

  return [];
}
```

**Comment by do-one-thing-to-well** [comment]
en， confusion,how to chose image on the `ios` platform ?

**Comment by AKAPUCH** [comment]
does webview_flutter_wkwebview also have call back method for file <input> tag?
I want to show same action in my app, so I also need `setOnShowFileSelector` for WKWebview, too...

if this is difficult or hard to do, can anyone help me prevent showing file select dialog in iOS?
after that, I can use postMessage method to call same feature method for android.

**Comment by samderlust** [comment]
I tried to pick multiple files but only the first file get uploaded. the rest are ignored by the webview

**Comment by stuartmorgan** [comment]
@samderlust  This issue is about updating the README. If you have an issue with the functionality itself, you should file a new issue with details.

**Comment by collinsyu** [comment]
return img path success ,while input onchange cannot be active

**Comment by bortan01** [comment]
I leave the complete solution here

import 'package:image/image.dart' as image;
import 'package:image_picker/image_picker.dart' as image_picker;
import 'package:webview_flutter_android/webview_flutter_android.dart'
    as webview_flutter_android;

class Example extends StatefulWidget {
  const Example({super.key});

  @override
  State<Example> createState() =>
      _ExampleState();
}

class _ExampleState
    extends State<Example> {
  
  late final WebViewController controller;

  @override
  void dispose() {
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    controller = WebViewController()
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setBackgroundColor(const Color(0x00000000))
      ..loadRequest(
        Uri.parse(
          'https://your-url.com',
        ),
      );
    addFilesController(controller);
  }

  Future<void> addFilesController(WebViewController controllerV) async {
    if (Platform.isAndroid) {
      var controller = (controllerV.platform
          as webview_flutter_android.AndroidWebViewController);
      await controller.setOnShowFileSelector(_androidFilePicker);
    }
  }

  Future<List<String>> _androidFilePicker(
    webview_flutter_android.FileSelectorParams params,
  ) async {
    if (params.acceptTypes.any((type) => type == 'image/*')) {
      var picker = image_picker.ImagePicker();
      var photo =
          await picker.pickImage(source: image_picker.ImageSource.gallery);

      if (photo == null) {
        return [];
      }

      var imageData = await photo.readAsBytes();
      var decodedImage = image.decodeImage(imageData)!;
      var scaledImage = image.copyResize(decodedImage, width: 500);
      var jpg = image.encodeJpg(scaledImage, quality: 90);

      var filePath = (await getTemporaryDirectory()).uri.resolve(
            './image_${DateTime.now().microsecondsSinceEpoch}.jpg',
          );
      var file = await File.fromUri(filePath).create(recursive: true);
      await file.writeAsBytes(jpg, flush: true);

      return [file.uri.toString()];
    }

    return [];
  }

  @override
  Widget build(BuildContext context) {
    context.read<EntrepreneurshipBloc>();
    return Scaffold(
      appBar: CustomSinglePageAppBar(
        title: Lang.locale().registerEntrepreneurship,
      ),
      body: SingleChildScrollView(
        child: SizedBox(
          height: MediaQuery.of(context).size.height, // O usa double.infinity
          child: WebViewWidget(
            controller: controller,
          ),
        ),
      ),
    );
  }
}

**Comment by rohankandwal** [comment]
There is some special settings required to setup file upload in Android. 

Create a function addFileSelectionListener and call it from initState.

```
void addFileSelectionListener() async {
    if (Platform.isAndroid) {
      final androidController = controller.platform as AndroidWebViewController;
      await androidController.setOnShowFileSelector(_androidFilePicker);
    }
  }
  
  Future<List<String>> _androidFilePicker(final FileSelectorParams params) async {
  final result = await FilePicker.platform.pickFiles();

  if (result != null && result.files.single.path != null) {
    final file = File(result.files.single.path!);
    return [file.uri.toString()];
  }
  return [];
}
```
I have create a detailed post on [how to perform file upload on Android in Flutter](https://theprogrammingway.com/flutter-file-upload-using-webview-on-android/ ). Check it out if you want more details.
